# Create metadata_csv and test_csv

This notebook implements zero-shot classification on chest X-ray images using pretrained models from TorchXRayVision.

## Setup and Imports

In [13]:
# Colab-specific setup (uncomment if using Colab)
from google.colab import drive
drive.mount('/content/drive')
%cd /content/drive/Shareddrives/CS231N/chestxray-classification

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).
/content/drive/Shareddrives/CS231N/chestxray-classification


In [14]:
# Install dependencies (uncomment if needed)
!pip install torchxrayvision python-dotenv scikit-learn matplotlib tqdm --quiet

In [15]:
# Standard imports
import os
import sys
import logging
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from tqdm.autonotebook import tqdm
import torch
import torchxrayvision as xrv
from PIL import Image


<ipython-input-15-6dc81de95d8e>:8: TqdmExperimentalWarning: Using `tqdm.autonotebook.tqdm` in notebook mode. Use `tqdm.tqdm` instead to force console mode (e.g. in jupyter console)
  from tqdm.autonotebook import tqdm


In [16]:
# Add the parent directory to path (if running in notebooks/)
import shutil

src = "/content/drive/Shareddrives/CS231N/chestxray-classification"
dst = "/content/chestxray-classification"

shutil.copytree(
    src,
    dst,
    ignore=shutil.ignore_patterns(
        "assignment4", ".git", "__pycache__", "*.ipynb_checkpoints"
    ),
    dirs_exist_ok=True
)
sys.path.insert(0, "/content/chestxray-classification")


In [18]:
# Import custom modules
from utils.config import Config
from models.xray_models_load import load_models

In [19]:
# Setup logging
logging.basicConfig(
    level=logging.INFO,
    format="%(asctime)s - %(levelname)s - %(name)s - %(message)s",
    datefmt="%Y-%m-%d %H:%M:%S",
    stream=sys.stdout,      # Redirect to notebook output
    force=True
)
logger = logging.getLogger("xray_evaluation")

## Configuration

In [20]:
# Check if paths are set correctly
print(f"Data path: {Config.data_path}")
print(f"Image path: {Config.image_path}")
print(f"CSV path: {Config.csv_path}")
print(f"Output path: {Config.output_path}")
print(f"Using device: {Config.device}")
print(f"Target labels: {Config.target_labels}")

Data path: /content/drive/Shareddrives/CS231N/assignment4/cs231n/datasets/nih-chestxray
Image path: /content/drive/Shareddrives/CS231N/assignment4/cs231n/datasets/nih-chestxray/images
CSV path: /content/drive/Shareddrives/CS231N/assignment4/cs231n/datasets/nih-chestxray/Data_Entry_2017_v2020.csv
Output path: results
Using device: cpu
Target labels: ['Cardiomegaly', 'Atelectasis', 'Effusion', 'Pneumothorax']


In [21]:
# Update paths if needed
# Config.data_path = "/path/to/data"
# Config.image_path = os.path.join(Config.data_path, "images")
# Config.csv_path = os.path.join(Config.data_path, "Data_Entry_2017_v2020.csv")

# Create output directory
os.makedirs(Config.output_path, exist_ok=True)

## Load Models

In [10]:
# List available models
print("Available pretrained models:")
for model_name in xrv.models.model_urls.keys():
    print(f"- {model_name}")

Available pretrained models:
- all
- densenet121-res224-all
- nih
- densenet121-res224-nih
- pc
- densenet121-res224-pc
- chex
- densenet121-res224-chex
- rsna
- densenet121-res224-rsna
- mimic_nb
- densenet121-res224-mimic_nb
- mimic_ch
- densenet121-res224-mimic_ch
- resnet50-res512-all


In [22]:
# Load models
model_mimic, model_chex, pathologies = load_models(Config.device)

2025-05-22 02:51:32 - INFO - models.xray_models_load - Loading DenseNet121 pretrained on MIMIC-CXR...
2025-05-22 02:51:33 - INFO - models.xray_models_load - Loading DenseNet121 pretrained on CheXpert...
2025-05-22 02:51:33 - INFO - models.xray_models_load - Models loaded successfully with 18 disease classes


In [23]:
# Display pathologies
print("\nSupported pathologies:")
for i, pathology in enumerate(pathologies):
    print(f"{i}: {pathology}")

# Highlight target pathologies
print("\nTarget pathologies:")
for label in Config.target_labels:
    idx = pathologies.index(label)
    print(f"{idx}: {label}")


Supported pathologies:
0: Atelectasis
1: Consolidation
2: 
3: Pneumothorax
4: Edema
5: 
6: 
7: Effusion
8: Pneumonia
9: 
10: Cardiomegaly
11: 
12: 
13: 
14: Lung Lesion
15: Fracture
16: Lung Opacity
17: Enlarged Cardiomediastinum

Target pathologies:
10: Cardiomegaly
0: Atelectasis
7: Effusion
3: Pneumothorax


## Load and Prepare Data

In [24]:
def encode_labels(row, all_labels):
    """
    Encode multi-label findings as binary vector

    Parameters:
    row (pandas.Series): DataFrame row containing 'Finding Labels'
    all_labels (list): List of all possible labels

    Returns:
    list: Binary encoding for each label
    """
    findings = row["Finding Labels"].split("|")
    return [int(label in findings) for label in all_labels]

def load_and_prepare_data(config, pathologies):
    """
    Load and prepare dataset with proper error handling

    Parameters:
    config (Config): Configuration object with paths and settings
    pathologies (list): List of all pathologies supported by models

    Returns:
    tuple: (test_df, label_indices)
    """
    try:
        logger.info(f"Loading metadata from {config.csv_path}")
        metadata_df = pd.read_csv(config.csv_path)
        logger.info(f"Raw metadata contains {len(metadata_df)} entries")

        # Get available images
        available_images = {f for f in os.listdir(config.image_path) if f.endswith(".png")}
        logger.info(f"Found {len(available_images)} PNG images in directory")

        # Filter to match available images
        metadata_df = metadata_df[metadata_df["Image Index"].isin(available_images)]
        logger.info(f"Filtered to {len(metadata_df)} entries with available images")


        # First encode the labels using your encode_labels function
        metadata_df["encoded_labels"] = metadata_df.apply(
            lambda row: encode_labels(row, pathologies), axis=1
        )
        # Then convert to numpy array
        metadata_df["encoded_array"] = metadata_df["encoded_labels"].apply(np.array)


        # Create label_indices as a dictionary
        label_indices = [pathologies.index(label) for label in config.target_labels]

        # Helper to check if all target diseases are negative
        def is_all_target_negative(row):
            return sum(row[label_indices]) == 0

        metadata_df["is_all_target_negative"] = metadata_df["encoded_array"].apply(
            lambda row: is_all_target_negative(row)
        )
        # Create separate columns using the list with target labels
        for disease, idx in zip(config.target_labels, label_indices):
          metadata_df[disease] = metadata_df["encoded_array"].apply(lambda x: x[idx])
        metadata_df.to_csv(os.path.join(config.output_path, "nih_metadata.csv"), index=False)
        return metadata_df, label_indices

    except Exception as e:
        logger.error(f"Failed to prepare data: {e}")
        raise

In [25]:
# Load and prepare data
metadata_df, label_indices = load_and_prepare_data(Config, pathologies)

2025-05-22 02:51:39 - INFO - xray_evaluation - Loading metadata from /content/drive/Shareddrives/CS231N/assignment4/cs231n/datasets/nih-chestxray/Data_Entry_2017_v2020.csv
2025-05-22 02:51:40 - INFO - xray_evaluation - Raw metadata contains 112120 entries
2025-05-22 02:51:40 - INFO - xray_evaluation - Found 54999 PNG images in directory
2025-05-22 02:51:41 - INFO - xray_evaluation - Filtered to 54999 entries with available images


In [26]:
!pip install python-dotenv
import os
from dotenv import load_dotenv

load_dotenv(".env")  # loads GH_TOKEN into os.environ

# Now safely retrieve the token
token = os.getenv("GH_TOKEN")
username = "havindh"
repo = "chestxray-classification"
push_url = f"https://{username}:{token}@github.com/{username}/{repo}.git"

In [27]:
!git status

On branch main
Your branch is ahead of 'origin/main' by 2 commits.
  (use "git push" to publish your local commits)

Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
	modified:   assignment4
	modified:   notebooks/create_metadata_csv.ipynb

no changes added to commit (use "git add" and/or "git commit -a")


In [28]:
metadata_df.columns

Index(['Image Index', 'Finding Labels', 'Follow-up #', 'Patient ID',
       'Patient Age', 'Patient Sex', 'View Position', 'OriginalImage[Width',
       'Height]', 'OriginalImagePixelSpacing[x', 'y]', 'encoded_labels',
       'encoded_array', 'is_all_target_negative', 'Cardiomegaly',
       'Atelectasis', 'Effusion', 'Pneumothorax'],
      dtype='object')

In [33]:
!git status

On branch main
Your branch is ahead of 'origin/main' by 2 commits.
  (use "git push" to publish your local commits)

Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
	modified:   assignment4
	modified:   notebooks/create_metadata_csv.ipynb

no changes added to commit (use "git add" and/or "git commit -a")


In [35]:
!git add notebooks/create_metadata_csv.ipynb
!git add results/nih_test_set_balanced.csv

In [38]:
!git push {push_url} main

Everything up-to-date


In [ ]:
metadata_df.head()

In [30]:
def get_test_set(metadata_df, label_indices, config):
  # Create test set with balanced classes
  test_df = create_balanced_test_set(metadata_df, label_indices, config)
  return test_df

def create_balanced_test_set(metadata_df, label_indices, config):
    """
    Create a balanced test set with same number of samples per disease

    Parameters:
    metadata_df (pandas.DataFrame): DataFrame with encoded labels
    label_indices (list): Indices of target labels
    config (Config): Configuration object with settings

    Returns:
    pandas.DataFrame: Balanced test set
    """
    # Helper to check if a disease is positive
    def is_positive(label_index):
        return metadata_df["encoded_array"].apply(lambda x: x[label_index] == 1)



    # Sample positive examples for each disease
    # Sample 400 positive per disease
    positive_samples = []

    for idx, label in zip(label_indices, config.target_labels):
        positives = metadata_df[is_positive(idx)]
        # Handle case where there might be fewer than 400 samples
        n_samples = min(400, len(positives))
        sampled = positives.sample(n=n_samples, random_state=42)
        positive_samples.append(sampled)

    # Combine all disease-positive samples
    positive_samples = pd.concat(positive_samples)

    # Sample negative examples (zero for all target diseases)
    negatives = metadata_df[metadata_df["is_all_target_negative"]]
    logger.info(f"Found {len(negatives)} cases negative for all target diseases")
    if len(negatives) < config.samples_per_disease:
        logger.warning(
            f"Only {len(negatives)} negative samples available, "
            f"wanted {config.samples_per_disease}"
        )
        negative_df = negatives
    else:
        negative_df = negatives.sample(n=config.samples_per_disease, random_state=config.random_seed)

    # Combine, deduplicate and prepare test set
    test_df = pd.concat([positive_samples, negative_df])
    test_df = test_df.drop_duplicates(subset="Image Index").reset_index(drop=True)
    logger.info(f"Created test set with {len(test_df)} unique samples after deduplication")

    # # Extract individual columns for each target disease
    # for i, label in zip(label_indices, config.target_labels):
    #     test_df[label] = test_df["encoded_array"].apply(lambda x: x[i])

    # Save test set for reproducibility
    os.makedirs(config.output_path, exist_ok=True)
    test_df[["Image Index"]].to_csv(os.path.join(config.output_path, "nih_test_set_balanced.csv"), index=False)

    return test_df

In [31]:
# Load and prepare test_data set
test_df = get_test_set(metadata_df,label_indices, Config)

2025-05-22 02:53:33 - INFO - xray_evaluation - Found 42511 cases negative for all target diseases
2025-05-22 02:53:33 - INFO - xray_evaluation - Created test set with 1969 unique samples after deduplication


In [ ]:
# Visualize class distribution
plt.figure(figsize=(10, 5))
test_df[Config.target_labels].sum().plot(kind="bar", color="blue")
plt.title("Positive Case Count per Disease")
plt.ylabel("Count")
plt.grid(axis='y', alpha=0.3)
plt.tight_layout()
plt.show()

## Visualize Sample Images

In [ ]:
# Function to display sample X-ray images
def display_sample_images(test_df, image_path, num_samples=3):
    """Display sample X-ray images for each disease category"""
    fig, axes = plt.subplots(len(Config.target_labels) + 1, num_samples,
                            figsize=(num_samples * 4, (len(Config.target_labels) + 1) * 4))

    # For each disease, show samples
    for i, disease in enumerate(Config.target_labels):
        disease_samples = test_df[test_df[disease] == 1].sample(num_samples)

        for j, (_, row) in enumerate(disease_samples.iterrows()):
            img_file = os.path.join(image_path, row["Image Index"])
            img = Image.open(img_file).convert('L')
            axes[i, j].imshow(img, cmap='gray')
            axes[i, j].set_title(f"{disease}\n{row['Image Index']}")
            axes[i, j].axis('off')

    # Show negative samples (no target disease)
    negative_samples = test_df[(test_df[Config.target_labels] == 0).all(axis=1)].sample(num_samples)

    for j, (_, row) in enumerate(negative_samples.iterrows()):
        img_file = os.path.join(image_path, row["Image Index"])
        img = Image.open(img_file).convert('L')
        axes[-1, j].imshow(img, cmap='gray')
        axes[-1, j].set_title(f"No Target Disease\n{row['Image Index']}")
        axes[-1, j].axis('off')

    plt.tight_layout()
    plt.show()

In [ ]:
# Display sample images
display_sample_images(test_df, Config.image_path, num_samples=3)

## Conculison

Metadata:  ???

Test dataset balanced

Found 42511 cases negative for all target diseases

Created test set with 1969 unique samples after deduplication